### ISSUE TITLE
Predicting on One Sentence whose Target is Not Known using Convolutional Seq2Seq

### ISSUE THREAD

**Issue by kuldeep7688** [bentrevett/pytorch-seq2seq#39]
Suppose we are translating from French to English. 

So in rnn based encoder-decoder model, we initialize trg sentence as string of <sos> tokens. 

Can you please explain how this will be executed in convolution based encoder-decoder module?

In rnn based model we feed the predicted token at time-step t-1  (th) to t (th) step in decoder to get the token at t (th) . I am little bit confused about how this will be done in convolution based decoder.

**Comment by bentrevett** [comment]
Yeah, the inference for these non-recurrent models is a bit different. 

I'll admit I haven't implemented inference for it yet but I believe the way it works is in a giant loop. In the first iteration of the loop, `trg` is just a sequence starting with an `<sos>` token and a few padding tokens(?). You pass the input sentence and this dummy target sentence into the model, then you see what it predicted for the token after the `<sos>` token (ignoring all other predictions afterward). Say it predicted "the". You then do the second iteration of the loop, however now your `trg` is `<sos> the`. You then feed those two into the model again, only caring about it what it predicts for after "the" - say "man". You then do a third iteration of the loop, feeding in `<sos> the man`. You keep doing this until the model outputs an `<eos>` token or you've reached some maximum number of tokens generated or some other stopping condition.

Let me know if that sounds sensible or not!

The only thing I am not 100% sure about is if the model will sometimes change it's earlier predictions when predicting future words. For example you could feed in "the man sat on the" and it might output "the woman sat on the floor". Do you just ignore the fact that "man" changed to "woman"? Or maybe it won't ever do this! 

I'm hoping to finish the conv seq2seq and the transformer notebooks without inference first, and then go back and write the inference code. Hopefully will be done by the end of this month (maybe)!

**Comment by kuldeep7688** [comment]
According to my intuition in Convolution Seq2Seq teacher forcing ratio is 1.0 while training and maybe that's why it is performing really well on test and validation because we have a trg sentence. Let me know if you think the same.

Yeah, I was also thinking about the giant loop for inferencing if there is no trg sentence present. 
I will try writing the code for inferencing using convolution Seq2Seq and analyze the outputs. I will update you if there exists a situation ("man" and "woman" example) like this. 

It will also help us to understand the model better and if we are not getting the desired results we can maybe go through their https://github.com/facebookresearch/fairseq . And may even contact the people who published the paper about this.

**Comment by bentrevett** [comment]
Yeah, this model will always have a teacher forcing ratio of 1.0 so the validation/test perplexity are lower than they should be. Running the tutorial 3 model with a constant teacher forcing ratio of 1.0 gives the following results:

```
Epoch: 01 | Time: 0m 55s
	Train Loss: 4.556 | Train PPL:  95.218
	 Val. Loss: 3.598 |  Val. PPL:  36.537
Epoch: 02 | Time: 0m 55s
	Train Loss: 3.314 | Train PPL:  27.492
	 Val. Loss: 2.838 |  Val. PPL:  17.080
Epoch: 03 | Time: 0m 55s
	Train Loss: 2.583 | Train PPL:  13.242
	 Val. Loss: 2.256 |  Val. PPL:   9.545
Epoch: 04 | Time: 0m 55s
	Train Loss: 2.041 | Train PPL:   7.700
	 Val. Loss: 1.948 |  Val. PPL:   7.012
Epoch: 05 | Time: 0m 54s
	Train Loss: 1.665 | Train PPL:   5.285
	 Val. Loss: 1.819 |  Val. PPL:   6.167
Epoch: 06 | Time: 0m 55s
	Train Loss: 1.381 | Train PPL:   3.977
	 Val. Loss: 1.757 |  Val. PPL:   5.796
Epoch: 07 | Time: 0m 55s
	Train Loss: 1.160 | Train PPL:   3.189
	 Val. Loss: 1.768 |  Val. PPL:   5.860
Epoch: 08 | Time: 0m 54s
	Train Loss: 0.990 | Train PPL:   2.691
	 Val. Loss: 1.763 |  Val. PPL:   5.832
Epoch: 09 | Time: 0m 55s
	Train Loss: 0.849 | Train PPL:   2.337
	 Val. Loss: 1.783 |  Val. PPL:   5.947
Epoch: 10 | Time: 0m 55s
	Train Loss: 0.735 | Train PPL:   2.086
	 Val. Loss: 1.817 |  Val. PPL:   6.152
```

So pretty similar overall. Ideally, I should be comparing BLEU scores instead of perplexity - but have not got around to implementing this yet - and these scores should be done without any teacher forcing - which I can do once inference for the conv. seq2seq model is ready.

**Comment by bentrevett** [comment]
@kuldeep7688 In case you are still interested, the convolutional sequence to sequence tutorial now has inference added.
